drivers/nec/display/all.cr

```diff
@@ -0,0 +1,410 @@
+
+    # Double check the input again!
+    # @input_double_check.cancel if @input_double_check
+    # @input_double_check = schedule.in(4.seconds) do
+    #   @input_double_check = nil
+    #   video_input
+    # end
+    # TODO: check if the below is equivalent to what's above
+    @input_double_check.try &.terminate
+    schedule.in(4.seconds) do
+      @input_double_check = nil
+      video_input.as(PlaceOS::Driver::Transport)
+    end
+
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+  end
+
+  enum Audio
+    Audio1        = 1
+    Audio2        = 2
+    Audio3        = 3
+    Hdmi          = 4
+    Tv            = 6
+    Display_port  = 7
+  end
+
+  def switch_audio(input : Audio)
+    @target_audio = input
+
+    data = OPERATIONS[:audio_input]
+    data += input.value.to_s(16).upcase.rjust(4, '0')
+
+    do_send("set parameter", data, name: "audio")
+    mute_status(20) # higher status than polling commands - lower than input switching
+    volume_status(20)
+
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+  end
+
+  def auto_adjust
+    data = OPERATIONS[:audio_setup]
+    data += "0001"
+    # TODO: find out if there is an equivalent for delay_on_receive
+    do_send("set parameter", data)#, delay_on_receive: 4.seconds)
+  end
+
+  def brightness(val : Int32)
+    data = OPERATIONS[:brightness_status]
+    data += val.clamp(0, 100).to_s(16).upcase.rjust(4, '0')
+
+    do_send("set_parameter", data, name: "brightness")
+    do_send("command", "0C", name: "brightness_save") # Save the settings
+  end
+
+  def contrast(val : Int32)
+    data = OPERATIONS[:contrast_status]
+    data += val.clamp(0, 100).to_s(16).upcase.rjust(4, '0')
+
+    do_send("set_parameter", data, name: "contrast")
+    do_send("command", "0C", name: "contrast_save") # Save the settings
+  end
+
+  def volume(val : Int32)
+    data = OPERATIONS[:volume_status]
+    data += val.clamp(0, 100).to_s(16).upcase.rjust(4, '0')
+
+    do_send("set_parameter", data, name: "volume_status")
+    do_send("command", "0C", name: "volume_save") # Save the settings
+    self[:audio_mute] = false # audio is unmuted when the volume is set
+  end
+
+  def mute_audio(state : Bool = true, index : Int32 | String = 0)
+    data = OPERATIONS[:mute_status]
+    data += state ? "0001" : "0000"
+
+    do_send("set_parameter", data)
+    logger.debug { "requested to update mute to #{state}" }
+  end
+
+  def unmute_audio
+    mute_audio(false)
+  end
+
+  # LCD Response code
+  def received(data, task)
+    hex = String.new(data)
+    # Check for valid response
+    if !check_checksum(data)
+      return task.try &.retry("-- NEC LCD, invalid response was: #{hex}")
+    end
+
+    logger.debug { "NEC LCD responded with hex #{hex}" }
+
+    command = MSG_TYPE[hex[4]]
+
+    case command # Check the MSG_TYPE (B, D or F)
+      when "command_reply"
+        # Power on and off
+        if hex[10..15] == "C203D6" # Means power comamnd
```

Comment:
It would be good to encode all of these 'magic numbers' within an enum or similar.

This driver looks to have left a little to be desired from it's original implementation. I'd recommend using the protocol manual as the main reference, and using that to help build abstractions that work well here as opposed to focusing on a line-by-line port of the ruby driver.

Reply:
thanks @KimBurgess for the to_s(16)/to_i(16) idea, that does help clean up things a bit
https://github.com/PlaceOS/drivers/pull/50/commits/d0146a1fbb04bdc787182cff378aae0231a6624b

Reply:
also this commit which kind of neatens things maybe
https://github.com/PlaceOS/drivers/pull/50/commits/65744ac9f4abf981d6116f16a4904ee43f586866